adder/ipfsadd/add.go

```diff
@@ -0,0 +1,443 @@
+	// patch it into the root
+	if path == "" {
+		path = node.Cid().String()
+	}
+
+	if pi, ok := node.(*posinfo.FilestoreNode); ok {
+		node = pi.Node
+	}
+
+	mr, err := adder.mfsRoot()
+	if err != nil {
+		return err
+	}
+	dir := gopath.Dir(path)
+	if dir != "." {
+		opts := mfs.MkdirOpts{
+			Mkparents: true,
+			Flush:     false,
+			Prefix:    adder.Prefix,
+		}
+		if err := mfs.Mkdir(mr, dir, opts); err != nil {
+			return err
+		}
+	}
+
+	if err := mfs.PutNode(mr, path, node); err != nil {
+		return err
+	}
+
+	if !adder.Silent {
+		return outputDagnode(adder.ctx, adder.Out, path, node)
+	}
+	return nil
+}
+
+// AddFile adds a file to the dagservice and outputs through the outchan
+func (adder *Adder) AddFile(file files.File) error {
+	if adder.liveNodes >= liveCacheSize {
+		// TODO: A smarter cache that uses some sort of lru cache with an eviction handler
+		mr, err := adder.mfsRoot()
+		if err != nil {
+			return err
+		}
+		if err := mr.FlushMemFree(adder.ctx); err != nil {
+			return err
+		}
+
+		adder.liveNodes = 0
+	}
+	adder.liveNodes++
+
+	if file.IsDirectory() {
+		return adder.addDir(file)
+	}
+
+	// case for symlink
+	if s, ok := file.(*files.Symlink); ok {
+		sdata, err := unixfs.SymlinkData(s.Target)
+		if err != nil {
+			return err
+		}
+
+		dagnode := dag.NodeWithData(sdata)
+		dagnode.SetPrefix(adder.Prefix)
+		err = adder.dagService.Add(adder.ctx, dagnode)
+		if err != nil {
+			return err
+		}
+
+		return adder.addNode(dagnode, s.FileName())
+	}
+
+	// case for regular file
+	// if the progress flag was specified, wrap the file so that we can send
+	// progress updates to the client (over the output channel)
+	var reader io.Reader = file
+	if adder.Progress {
+		rdr := &progressReader{file: file, out: adder.Out}
+		if fi, ok := file.(files.FileInfo); ok {
+			reader = &progressReader2{rdr, fi}
+		} else {
+			reader = rdr
+		}
+	}
+
+	dagnode, err := adder.add(reader)
+	if err != nil {
+		return err
+	}
+
+	// patch it into the root
+	return adder.addNode(dagnode, file.FileName())
+}
+
+func (adder *Adder) addDir(dir files.File) error {
+	log.Infof("adding directory: %s", dir.FileName())
+
+	mr, err := adder.mfsRoot()
+	if err != nil {
+		return err
+	}
+	err = mfs.Mkdir(mr, dir.FileName(), mfs.MkdirOpts{
+		Mkparents: true,
+		Flush:     false,
+		Prefix:    adder.Prefix,
+	})
+	if err != nil {
+		return err
+	}
+
+	for {
+		file, err := dir.NextFile()
+		if err != nil && err != io.EOF {
+			return err
+		}
+		if file == nil {
+			break
+		}
+
+		// Skip hidden files when adding recursively, unless Hidden is enabled.
+		if files.IsHidden(file) && !adder.Hidden {
+			log.Infof("%s is hidden, skipping", file.FileName())
+			continue
+		}
+		err = adder.AddFile(file)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// outputDagnode sends dagnode info over the output channel
+func outputDagnode(ctx context.Context, out chan *api.AddedOutput, name string, dn ipld.Node) error {
```

Comment:
s/outputDagnode/outputDagNode or outputDAGNode whichever is fine

Reply:
hmm this code is from go-ipfs I'm not sure we should touch more than necessary..

Reply:
not so much. PR it to go-ipfs first :) I need the cleanest diff possible for the moment.